sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java

```diff
@@ -240,63 +245,109 @@ private Read(
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      Read read = (Read) o;
+      return configuration.toString().equals(read.configuration.toString())
+          && Objects.equals(tableId, read.tableId)
+          && scan.toString().equals(read.scan.toString());
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(configuration, tableId, scan);
+    }
+
+    private Object writeReplace() {
+      return new SerializationProxy(this);
+    }
+
+    private static class SerializationProxy implements Serializable {
+      public SerializationProxy() {}
+
+      public SerializationProxy(Read read) {
+        configuration = read.configuration;
+        tableId = read.tableId;
+        scan = read.scan;
+      }
+
+      private void writeObject(ObjectOutputStream out) throws IOException {
+        SerializableCoder.of(SerializableConfiguration.class)
+            .encode(new SerializableConfiguration(this.configuration), out);
+        StringUtf8Coder.of().encode(this.tableId, out);
+        ProtobufUtil.toScan(this.scan).writeDelimitedTo(out);
+      }
+
+      private void readObject(ObjectInputStream in) throws IOException {
+        this.configuration = SerializableCoder.of(SerializableConfiguration.class).decode(in).get();
+        this.tableId = StringUtf8Coder.of().decode(in);
+        this.scan = ProtobufUtil.toScan(ClientProtos.Scan.parseDelimitedFrom(in));
+      }
+
+      Object readResolve() {
+        return HBaseIO.read().withConfiguration(configuration).withTableId(tableId).withScan(scan);
+      }
+
+      private Configuration configuration;
+      private String tableId;
+      private Scan scan;
+    }
+
+    @SuppressFBWarnings("SE_BAD_FIELD")
+    private final Configuration configuration;
+
     private final String tableId;
-    private final SerializableScan serializableScan;
+
+    @SuppressFBWarnings("SE_BAD_FIELD")
+    private final Scan scan;
   }
 
   /**
    * A {@link PTransform} that works like {@link #read}, but executes read operations coming from a
-   * {@link PCollection} of {@link HBaseQuery}.
+   * {@link PCollection} of {@link Read}.
    */
   public static ReadAll readAll() {
-    return new ReadAll(null);
+    return new ReadAll();
   }
 
   /** Implementation of {@link #readAll}. */
-  public static class ReadAll extends PTransform<PCollection<HBaseQuery>, PCollection<Result>> {
-
-    private ReadAll(SerializableConfiguration serializableConfiguration) {
-      this.serializableConfiguration = serializableConfiguration;
-    }
-
-    /** Reads from the HBase instance indicated by the* given configuration. */
-    public ReadAll withConfiguration(Configuration configuration) {
-      checkArgument(configuration != null, "configuration can not be null");
-      return new ReadAll(new SerializableConfiguration(configuration));
-    }
+  public static class ReadAll extends PTransform<PCollection<Read>, PCollection<Result>> {
```

Comment:
Is it a breaking change of users API?

Reply:
Yes and it is intended, notice however that the only breaking change is for the SDF based read transform `ReadAll`. Also let's not forget that HBaseIO is still Experimental and Reads based on SDF for HBaseIO are Experimental^2. However after this change I seriously doubt the API will change in the future.

Reply:
Please, add this warning/note to top class Javadoc

Reply:
I better do this in the CHANGES.md release notes file, so this gets announced with the release notes. Java will cover making users aware at the code level :)